cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -517,36 +521,52 @@ public void upgradeApplications(FullHttpRequest request, HttpResponder responder
                                   @PathParam("namespace-id") String namespaceId,
                                   @QueryParam("artifactScope") Set<String> artifactScopes,
                                   @QueryParam("allowSnapshot") boolean allowSnapshot) throws Exception {
-    // TODO: (CDAP-16910) Improve batch API performance as each application upgrade is an event independent of each
-    //  other.
-
     List<ApplicationId> appIds = decodeAndValidateBatchApplicationRecord(validateNamespace(namespaceId), request);
     Set<ArtifactScope> allowedArtifactScopes = getArtifactScopes(artifactScopes);
     try (ChunkResponder chunkResponder = responder.sendChunkStart(HttpResponseStatus.OK)) {
       ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
       try (JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8))) {
         jsonWriter.beginArray();
+        List<Future<?>> futureList = new ArrayList<>();
+        final ReentrantLock lock = new ReentrantLock();
+        ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(10);
```

Comment:
Don't use casting. If you want a `ThreadPoolExecutor`, new it directly. However, in this case, it just need to be an `ExecutorService`, which is what returned by the newFixedThreadPool method.

Also, when creating executor, always provide a `ThreadFactory` to name threads it created. It helps stacktrace analysis much easier (and consider using daemon thread in this case).

One last thing, I don't see this ExecutorService is being terminated after use, which can lead to thread leakage.

Reply:
Done. I have used the defaultThreadFactory btw, do we want a customized threadpoolfactory?